### ISSUE TITLE
Find a way to avoid bad relevance incompleteness in conversion

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#12809]
cf https://coq.github.io/doc/master/refman/addendum/sprop.html#issues-with-non-cumulativity

<details>
```coq
(* -*- coq-prog-args: ("-allow-sprop") -*- *)
Inductive sFalse : SProp := .
Inductive sTrue : SProp := sI.
Inductive seq {A} (x : A) : A -> SProp := seq_refl : seq x x.
Inductive sseq {A : SProp} (x : A) : A -> SProp := sseq_refl : sseq x x.
Declare Scope sprop_scope.
Delimit Scope sprop_scope with sprop.
Infix "=" := seq : sprop_scope.
Arguments seq {A} x _.
Arguments seq_refl {A x} , [A] x.
Arguments sseq {A} x _.
Arguments sseq_refl {A x} , [A] x.

Definition false_nseq_true (H : seq false true) : sFalse
  := match H in seq _ y return if y return SProp then sFalse else sTrue with
     | seq_refl => sI
     end.

Definition true_nseq_false (H : seq true false) : sFalse
  := match H in seq _ y return if y return SProp then sTrue else sFalse with
     | seq_refl => sI
     end.

Definition not_sFalse (p : sFalse) : False := match p with end.

Definition seq_bool_rect (x : bool)
           (P : forall y, seq x y -> Type)
           (h : P x seq_refl)
  : forall y p, P y p.
Proof.
  destruct x, y; intro p.
  refine (_ : P true seq_refl).
  (* Error:
Ltac call to "refine (uconstr)" failed.
In environment
P : forall y : bool, (true = y)%sprop -> Type
h : P true seq_refl
p : (true = true)%sprop
The term "?p : P true seq_refl" has type "P true seq_refl"
while it is expected to have type "P true p".
*)
```
This seems like a bug.
</details>
#### Coq Version

f238b252019555c0a9ae5bb0f56b4c73192b1e86

**Comment by JasonGross** [comment]
And, indeed, I can give the term by hand...
```coq
Definition seq_bool_rect (x : bool)
           (P : forall y, seq x y -> Type)
           (h : P x seq_refl)
           y
  : forall p, P y p
  := match x, y return forall (P : forall y', seq x y' -> Type)
                              (h : P x seq_refl)
                              p, P y p with
     | true, true => fun P h p => h
     | false, false => fun P h p => h
     | true, false => fun P _ pf => match true_nseq_false pf return P _ pf with end
     | false, true => fun P _ pf => match false_nseq_true pf return P _ pf with end
     end P h.
```

**Comment by ppedrot** [comment]
It's a know problem about relevance annotations not being correctly set. cc @SkySkimmer, FWIW.

**Comment by SkySkimmer** [comment]
Yes, the relevance on `p` is incorrect because its type is implicit. For instance if you do
~~~coq
Definition seq_bool_rect (x : bool)
           (P : forall y, seq x y -> Type)
           (h : P x seq_refl)
  : forall y (p : _ : SProp), P y p.
~~~
it works.

**Comment by JasonGross** [comment]
Why did you close this?  It seems like clearly a bug to me, even if it's not a bug in conversion and is instead a bug in elaboration, and even if it's hard to solve.

**Comment by ppedrot** [comment]
Fixed by #17091.
